### PR TITLE
powerpc plan for linux_debug.c

### PR DESCRIPTION
**Pull request by unixfreaxjp**
Ref: 
https://github.com/radare/radare2/issues/4720#issuecomment-218679045
https://github.com/radare/radare2/issues/4720#issuecomment-218731890
https://github.com/radare/radare2/issues/4873

### CODE REVIEW

**Comment by XVilka** [comment]
Surely inspired by this PR https://github.com/radare/radare2/pull/4864

File: libr/debug/p/native/linux/linux_debug.c [hunk 1]
```diff
@@ -203,6 +203,8 @@ RList *linux_thread_list (int pid, RList *list) {
 	eprintf ("fos = 0x%04lx              ", (fpregs).fos)
 
 static void print_fpu (void *f, int r){
+#if __ppc__ || __powerpc__ || __POWERPC__ // added ppc
+	eprintf ("---- powerpc ----\n");
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: useless debug printf

File: libr/debug/p/native/linux/linux_debug.c [hunk 2]
```diff
@@ -395,7 +395,7 @@ int linux_reg_read (RDebug *dbg, int type, ut8 *buf, int size) {
 			};
 			ret = ptrace (PTRACE_GETREGSET, pid, NT_PRSTATUS, &io);
 			}
-#elif __POWERPC__
+#elif __POWERPC__ || __powerpc__ || __ppc__ // same ret for ppc32
 			ret = ptrace (PTRACE_GETREGS, pid, &regs, NULL);
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: this is wrong.  powerpc's ptrace swaps last two arguments, see manpage

File: libr/debug/p/native/linux/linux_debug.c [hunk 3]
```diff
@@ -203,20 +203,20 @@ RList *linux_thread_list (int pid, RList *list) {
...
-		ut64 *b = (ut64 *)(&fpregs.st_space[i*4]);
-		ut32 *c = (ut32*)&fpregs.st_space;
-		float *f = (float *)&fpregs.st_space;
-		c = c + (i * 4);
-		f = f + (i * 4);
-		eprintf ("st%d =%0.3lg (0x%016"PFMT64x") | %0.3f (%08x)  | \
-			%0.3f (%08x) \n", i,
-			(double)*((double*)&fpregs.st_space[i*4]), *b, (float) f[0],
-			c[0], (float) f[1], c[1]);
+#if __ppc__ || __powerpc__                              // reserved powepc;  tofix
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: do not indent comments

**Comment by radare** [comment]
can you squash? its pretty hard to review the changes this way

**Comment by unixfreaxjp** [comment]
what is "squash" means?

File: libr/debug/p/native/linux/linux_debug.c (hunk 1 above)
Comment (unixfreaxjp) [inline libr/debug/p/native/linux/linux_debug.c]: I used it to check whether the flag hits, want me to delete it now?

**Comment by unixfreaxjp** [comment]
squashed..

**Comment by unixfreaxjp** [comment]
I am sorry for adding https://github.com/radare/radare2/pull/4870/commits/1a4418f1effeeb7ec5cf3002ee30a5b57d6d850a
Issue: https://github.com/radare/radare2/issues/4873 
PoC: https://www.youtube.com/edit?video_id=iczDryFBb0Q

File: libr/debug/p/native/linux/linux_debug.c (hunk 2 above)
Comment (unixfreaxjp) [inline libr/debug/p/native/linux/linux_debug.c]: > powerpc's ptrace swaps last two arguments

you're right, thank you.

File: configure [hunk 4]
```diff
@@ -551,5 +551,10 @@ eval VAL="\$${A}"
 [ -z "${VAL}" ] && VAL="(null)"
 echo " - ${A} = ${VAL}"
 done
+
+# This will not return correct values (tested):
```
Comment (radare) [inline configure]: this change is doing nothing, and it is autogenerated, so it shouldnt be changed

File: configure [hunk 5]
```diff
@@ -551,3 +551,5 @@ eval VAL="\$${A}"
 [ -z "${VAL}" ] && VAL="(null)"
 echo " - ${A} = ${VAL}"
 done
+echo " - ARCH = `uname -m|sed -e 's, ,,g'|cut -d - -f 1`"
```
Comment (radare) [inline configure]: modify configure.acr, not the configure one, because this file is autogenerated

File: configure [hunk 6]
```diff
@@ -552,9 +552,16 @@ eval VAL="\$${A}"
...
 # echo " - ARCH = ${CPU}"
 # echo " - OSFAMILY = ${OS}"
 # Corrected (tested):
-echo " - ARCH = `uname -m|sed -e 's, ,,g'|cut -d - -f 1`"
-echo " - OSFAMILY = `uname -s|tr A-Z a-z`"
+# echo " - ARCH = `uname -m|sed -e 's, ,,g'|cut -d - -f 1`"
+# echo " - OSFAMILY = `uname -s|tr A-Z a-z`"
+
+# Beautified, refreshed shifted value for the need vars only:
+CPU="`uname -m|sed -e 's, ,,g'|cut -d - -f 1`"
```
Comment (radare) [inline configure]: dont do this. change acr, or configure.acr if needed. i will evaluate the changes there

**Comment by radare** [comment]
please squash

**Comment by unixfreaxjp** [comment]
> dont do this. change acr, or configure.acr if needed. i will evaluate the changes there

1. okay. putting this changes to configure.acr
2. squashing after configure.acr is done and will pull req again afterward.

**Comment by unixfreaxjp** [comment]
>  okay. putting this changes to configure.acr

[configure.acr](https://github.com/radare/radare2/blob/master/configure.acr) is not the source of this bug, escalating to [acr](https://github.com/radare/acr).

**Comment by unixfreaxjp** [comment]
## Proposing chages in `acr/src/acr-sh` code:

> configure.acr is not the source of this bug, escalating to acr.
> Reading the src/acr-sh.

The below `configure `codes:
```sh
do_remove
echo
echo "Final report:"
for A in  PREFIX HAVE_LIB_GMP HAVE_OPENSSL USE_CAPSTONE HAVE_FORK VERSION USE_LIB_ZIP USE_LIB_MAGIC DEBUGGER CC USERCC USEROSTYPE LIBVERSION ; do # REPORT
eval VAL="\$${A}"
[ -z "${VAL}" ] && VAL="(null)"
echo " - ${A} = ${VAL}"
done
```
was generated by `acr/src/acr-sh` in L:1781:
```sh
echo
echo "Final report:"
for A in ${REPORT} ; do # REPORT
eval VAL="\\${S}${S}{A}"
[ -z "${S}{VAL}" ] && VAL="(null)"
echo " - ${S}{A} = ${S}{VAL}"
done
EOF
```
As you may see the "enhancement" is proposing addition for "Final Report" to be more in detail with ARCH & OSFAMILY to support value USEROSTYPE  which is too generic, by adding the below marked ones:
```asm
Final report:
 - PREFIX = /usr/local
 - HAVE_LIB_GMP = 0
 - HAVE_OPENSSL = 0
 - USE_CAPSTONE = 0
 - HAVE_FORK = 1
 - VERSION = 0.10.3-git
 - USE_LIB_ZIP = 0
 - USE_LIB_MAGIC = 0
 - DEBUGGER = 1
 - CC = gcc
 - USERCC = gcc
 - USEROSTYPE = gnulinux
 - LIBVERSION = 0.10.3-git
 - ARCH = armv5tejl   /* <==== this  */
 - OSFAMILY = linux    /* <=== this */

(or in AMD64 FreebSD:)
Final report:
 - PREFIX = /usr/local
 - HAVE_LIB_GMP = 0
 - HAVE_OPENSSL = 0
 - USE_CAPSTONE = 0
 - HAVE_FORK = 1
 - VERSION = 0.10.3-git
 - USE_LIB_ZIP = 0
 - USE_LIB_MAGIC = 0
 - DEBUGGER = 1
 - CC = gcc
 - USERCC = gcc
 - USEROSTYPE = bsd
 - LIBVERSION = 0.10.3-git
 - ARCH = amd64  /* <==== this  */
 - OSFAMILY = freebsd   /* <==== this  */
```
So seeing the arc/src/arc-sh code pasted above it seem that `${REPORT}` is depending to the parameter inputted by configure request (with --with-xxxx) or auto, and yet, the `ARCH `and `OSFAMILY `is needed only by the auto value, shortly, with the current acr-sh is not extract-able by `${REPORT}` .

Therefore if team & pancake agree to add  `ARCH `and `OSFAMILY`, this can be done in two ways:
1. plastic surgery the `acr/src/acr-sh` to support `ARCH `and `OSFAMILY` in `${REPORT}` so we don't have to change the `acr/src/acr-sh` in L:1781 or ....
2. just added below code in `acr/src/acr-sh` in L:1788 between `done` and `EOF`:
```sh
CPU="`uname -m|sed -e 's, ,,g'|cut -d - -f 1`"
OS="`uname -s|tr A-Z a-z`"
echo " - ARCH = ${CPU}"
echo " - OSFAMILY = ${OS}"
echo # added ¥n need to mark end Final Report
```
If the addition in src/acr-sh is made, either option 1 or 2 above, the configure additional code is removed. 

So, as per advised:

> dont do this. change acr, or configure.acr if needed. i will evaluate the changes there

I can not change with option one,without pancake territory to judge first. So I go with the option 2. For the pullreq of this modification, conclusion:
1. I go with option 2
2. I made pullreq
3. pancake will decide to do the merge or to recode src/arc-sh with the option 1.

Thank you/unixfreaxjp

**Comment by radare** [comment]
Make a pr in the acr repo for proper review

**Comment by radare** [comment]
And you dint need to modify the src of acr to do this.bjust add those vars in the reportline of the configure.acr

**Comment by unixfreaxjp** [comment]
> And you dint need to modify the src of acr to do this.bjust add those vars in the reportline of the configure.acr

Thank you for the suggestion.
In that case the code will be added to `configure.acr` [L:138](https://github.com/unixfreaxjp/radare2/blob/master/configure.acr#L138) instead of `acr/src/acr-sh` [L:1788](https://github.com/unixfreaxjp/acr/blob/master/src/acr-sh#L1788). 
Pulling request shortly.

**Comment by unixfreaxjp** [comment]
> And you dint need to modify the src of acr to do this.bjust add those vars in the reportline of the configure.acr

Negative. **Just [tested](https://github.com/unixfreaxjp/radare2/blob/master/configure.acr#L137)**. The variables for `${OS} `and `${CPU} `were not defined can not be extracted just by adding reportline in `configure.acr`. Please read long explanation in https://github.com/radare/radare2/pull/4870#issuecomment-219269732 
The changed was made in

**Comment by unixfreaxjp** [comment]
Executed pull request.

**Comment by radare** [comment]
The configure stuff is already in master and the debug one contains no changes, and the commit message is wrong (many 'update ..')

**Comment by unixfreaxjp** [comment]
> debug one contains no changes

Thank's for opportunity to consider this architecture specifically.
To be clear, currently there's no specific `__ppc__ || __powerpc__` compatibility for Cisco routers _PPC7* cpu_ or _powermac3g PPC_ (i.e. Blue/White series or Beige w/linux).
I had to delete codes in `print_fpu()` (not confidence struct for `fpregs`), `linux_reg_read()` and `linux_reg_write() `after trial & error, hoping current flow may cover PPC (doubtful). 
For the future error occurred, improvement in those parts maybe needed. Will add build test under this msg.

日本の皆様、残念ですがppcの対応が未だ不可能です。自分のコードをリファレンス不足で削除しなければなりません。対応の為に`print_fpu()`のfpregsを書く事と`linux_reg_read()` や `linux_reg_write()`が必要。フラグのトラッピングは`__ppc__ || __powerpc__`で確認が出来ましたので、使って下さい。尚、最新版ppcのビルドテストをこのメッセージ後に書きます。
以上/unixfreaxjp

**Comment by unixfreaxjp** [comment]
The MACRO used for ppc stated in above was incorrect. The correct and tested success ones for Debian distributions are `__powerpc__` or `__powerpc` for the arch 32bit and `__powerpc64__` for 64bit. Hence the Debian Linux/PPC print_fpu and registers are in the `/usr/include/powerpc-linux-gnu/`.
@radare is the one who figured this out, the commit that fixes this matter is in https://github.com/radare/radare2/commit/bc0933db54d2a698f56e5e284579578402313cc6 with the screenshots showing compilation success. The next steps are some adjustment in the register readings.

**Pull request closed without merging**